src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SslOverTdsStream.NetCoreApp.cs

```diff
@@ -0,0 +1,305 @@
+        {
+            return ReadAsync(new Memory<byte>(buffer, offset, count), cancellationToken).AsTask();
+        }
+
+        public override Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            return WriteAsync(new ReadOnlyMemory<byte>(buffer, offset, count), cancellationToken).AsTask();
+        }
+
+        public override int Read(Span<byte> buffer)
+        {
+            if (!_encapsulate)
+            {
+                return _stream.Read(buffer);
+            }
+
+            using (SNIEventScope.Create("<sc.SNI.SslOverTdsStream.Read |SNI|INFO|SCOPE> reading encapsulated bytes"))
+            {
+                if (_packetBytes > 0)
+                {
+                    // there are queued bytes from a previous packet available
+                    // work out how many of the remaining bytes we can consume
+                    int wantedCount = Math.Min(buffer.Length, _packetBytes);
+                    int readCount = _stream.Read(buffer.Slice(0, wantedCount));
+                    if (readCount == 0)
+                    {
+                        // 0 means the connection was closed, tell the caller
+                        return 0;
+                    }
+                    _packetBytes -= readCount;
+                    return readCount;
+                }
+                else
+                {
+                    Span<byte> headerBytes = stackalloc byte[TdsEnums.HEADER_LEN];
+
+                    // fetch the packet header to determine how long the packet is
+                    int headerBytesRead = 0;
+                    do
+                    {
+                        int headerBytesReadIteration = _stream.Read(headerBytes.Slice(headerBytesRead, TdsEnums.HEADER_LEN - headerBytesRead));
+                        if (headerBytesReadIteration == 0)
+                        {
+                            // 0 means the connection was closed, tell the caller
+                            return 0;
+                        }
+                        headerBytesRead += headerBytesReadIteration;
+                    } while (headerBytesRead < TdsEnums.HEADER_LEN);
+
+                    // read the packet data size from the header and store it in case it is needed for a subsequent call
+                    _packetBytes = ((headerBytes[TdsEnums.HEADER_LEN_FIELD_OFFSET] << 8) | headerBytes[TdsEnums.HEADER_LEN_FIELD_OFFSET + 1]) - TdsEnums.HEADER_LEN;
+
+                    // read as much from the packet as the caller can accept
+                    int packetBytesRead = _stream.Read(buffer.Slice(0, Math.Min(buffer.Length, _packetBytes)));
+                    _packetBytes -= packetBytesRead;
+                    return packetBytesRead;
+                }
+            }
+        }
+
+        public override async ValueTask<int> ReadAsync(Memory<byte> buffer, CancellationToken cancellationToken = default)
```

Comment:
If I am not mistaken ValueTask has some downsides as well. According to[ this article](https://devblogs.microsoft.com/dotnet/understanding-the-whys-whats-and-whens-of-valuetask/) from Stephen Toub:
```

I think using Task<TResult> is a better option.
However, because ValueTask and ValueTask<TResult> may wrap reusable objects, there are actually significant constraints on their consumption when compared with Task and Task<TResult>, should someone veer off the desired path of just awaiting them. In general, the following operations should never be performed on a ValueTask / ValueTask<TResult>:

Awaiting a ValueTask / ValueTask<TResult> multiple times. The underlying object may have been recycled already and be in use by another operation. In contrast, a Task / Task<TResult> will never transition from a complete to incomplete state, so you can await it as many times as you need to, and will always get the same answer every time.
Awaiting a ValueTask / ValueTask<TResult> concurrently. The underlying object expects to work with only a single callback from a single consumer at a time, and attempting to await it concurrently could easily introduce race conditions and subtle program errors. It’s also just a more specific case of the above bad operation: “awaiting a ValueTask / ValueTask<TResult> multiple times.” In contrast, Task / Task<TResult> do support any number of concurrent awaits.
Using .GetAwaiter().GetResult() when the operation hasn’t yet completed. The IValueTaskSource / IValueTaskSource<TResult> implementation need not support blocking until the operation completes, and likely doesn’t, so such an operation is inherently a race condition and is unlikely to behave the way the caller intends. In contrast, Task / Task<TResult> do enable this, blocking the caller until the task completes.
If you have a ValueTask or a ValueTask<TResult> and you need to do one of these things, you should use .AsTask() to get a Task / Task<TResult> and then operate on that resulting task object. After that point, you should never interact with that ValueTask / ValueTask<TResult> again.

The short rule is this: with a ValueTask or a ValueTask<TResult>, you should either await it directly (optionally with .ConfigureAwait(false)) or call AsTask() on it directly, and then never use it again, e.g.
```

Reply:
@Wraith2 any update here?

Reply:
No. Not really. I've followed the development of ValueTask in runtime, listened to the api reviews where Stephen Toub has explained the pitfalls and uses of them, read the blog posts. This PR contained what i consider a reasonable use of them enabling fast sync paths to be used in future. If you're not convinced there's not a lot i can do to convince you. @benadams had the same problem in https://github.com/dotnet/SqlClient/pull/335 , people just seem very hesitant on ValueTask and if he can't get it accepted very few people can.

Reply:
As also mentioned in https://github.com/dotnet/SqlClient/pull/541#issuecomment-656382148 since Async APIs are not called/referenced from the driver yet, this can be passed.